### PR TITLE
Consolidate Legato AF integration

### LINKED ISSUE #20: Integrate to legato runtime
build ta-endpoint as legato app

### PR DESCRIPTION
**Pull request by splasky**
These commits do the following things:
1. Fix warning during compiling.
2. Fix snprintf format on different platform.
3. Consolidate Legato AF intergration.

close #20

### CODE REVIEW

**Comment by jserv** [comment]
Use precise wording. Tweak the git commit messages accordingly.

File: Makefile [hunk 1]
```diff
@@ -51,27 +52,36 @@ mbedtls_make:
...
+http_parser.o: $(THIRD_PARTY_PATH)/http-parser/http_parser.c $(THIRD_PARTY_PATH)/http-parser/http_parser.h
+	$(CC) -c $(CFLAGS) $(INCLUDES) -o $@ $<
 main.o: main.c $(UTILS_OBJS)
 	@echo Compiling: $< ....
 	$(CC) -c $(CFLAGS) $(INCLUDES) -o $@ $<
 $(UTILS_OBJS):
 	$(MAKE) -C $(UTILS_PATH)
 test: $(TEST_PATH) $(UTILS_OBJS)
 	$(MAKE) -C $(TEST_PATH)
+ta_app: $(TALIB)
```
Comment (howjmay) [inline Makefile]: May you change it to `ta_leagto_app` here, since I think we should state clearly that this rule would generate legato app only.

File: utils/utils.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2018-2019 BiiLabs Co., Ltd. and Contributors
```
Comment (howjmay) [inline utils/utils.h]: 2019-2020

File: utils/utils.c [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "conn_http.h"
+#include "http_parser.h"
+
+static char addr_log_template[] = "\n%s\n";
+
+void gen_trytes(uint16_t len, char *out) {
```
Comment (howjmay) [inline utils/utils.c]: May you help me to rename it as `gen_rand_trytes()`

**Review by howjmay — CHANGES_REQUESTED** [review]

File: utils/crypto_utils.h [hunk 4]
```diff
@@ -19,12 +19,12 @@ extern "C" {
...
-int encrypt(unsigned char *plaintext, int plaintext_len, unsigned char *ciphertext, int ciphertext_len, uint8_t iv[16],
-            uint8_t key[AES_BLOCK_SIZE * 2], uint8_t device_id[IMSI_LEN + 1]);
-int decrypt(unsigned char *ciphertext, int ciphertext_len, unsigned char *plaintext, int plaintext_len, uint8_t iv[16],
-            uint8_t key[AES_BLOCK_SIZE * 2]);
+int get_device_id(char *device_id);
+int get_aes_key(uint8_t *key);
+int ta_encrypt(unsigned char *plaintext, int plaintext_len, unsigned char *ciphertext, int ciphertext_len,
+               uint8_t iv[16], uint8_t key[AES_BLOCK_SIZE * 2], uint8_t device_id[IMSI_LEN + 1]);
+int ta_decrypt(unsigned char *ciphertext, int ciphertext_len, unsigned char *plaintext, int plaintext_len,
+               uint8_t iv[16], uint8_t key[AES_BLOCK_SIZE * 2]);
```
Comment (howjmay) [inline utils/crypto_utils.h]: May you assign the size of IV with macro `AES_IV_SIZE` and define one more macro for `AES_KEY_SIZE`?
Btw are we using aes cbc here? If we are using cbc, then I think it is better to name the macro of iv size as `CBC_IV_SIZE`

**Review by howjmay — CHANGES_REQUESTED** [review]

File: endpointComponent/endpoint.c [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      if (ciphertext_len == 0) {
+        LE_ERROR("%s\n", "ta_encrypt msg error");
+      }
+      serialize_msg(iv, ciphertext_len, ciphertext, msg, &msg_len);
+      bytes_to_trytes((const unsigned char *)msg, msg_len, tryte_msg);
+
+      // Init http session. verify: check the server CA cert.
+      char msg_body[1024];
+      gen_tryte_msg(tryte_msg, addr, msg_body);
+      if (send_https_msg(HOST, PORT, API, msg_body, 1024) != HTTP_OK) {
```
Comment (howjmay) [inline endpointComponent/endpoint.c]: We should allow user to use other `host` and `port` here

**Review by howjmay — CHANGES_REQUESTED** [review]

File: endpointComponent/endpoint.c (hunk 5 above)
Comment (splasky) [reply endpointComponent/endpoint.c]: Enter by the command line or config file?

Comment (howjmay) [reply endpointComponent/endpoint.c]: How about using cli at the first step? 
I think it may be easier to implement

File: Makefile [hunk 6]
```diff
@@ -29,12 +29,13 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
...
 
-UTILS_OBJS = $(UTILS_PATH)/crypto_utils.o $(UTILS_PATH)/serializer.o $(UTILS_PATH)/tryte_byte_conv.o $(UTILS_PATH)/uart_utils.o
+UTILS_OBJS = $(UTILS_PATH)/crypto_utils.o $(UTILS_PATH)/serializer.o $(UTILS_PATH)/tryte_byte_conv.o $(UTILS_PATH)/uart_utils.o $(UTILS_PATH)/utils.o
 # We need to modify this rule here to be compatible to the situation 
 # that we have several different ways of connectivity in the future
 CONNECTIVITY_OBJS = conn_http.o
 
-OBJS = main.o $(HTTP_PARSER_PATH)/http_parser.o $(UTILS_OBJS) $(CONNECTIVITY_OBJS)
+OBJS = main.o http_parser.o $(UTILS_OBJS) $(CONNECTIVITY_OBJS)
+TALIB = libta_endpoint.so
```
Comment (jserv) [inline Makefile]: Change it to `TA_LIB`.

File: utils/utils.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+ * @param api http api
+ * @param msg message to send
+ * @param msg_len length of message array
+ *
+ * @return http status
+ * @see http_retcode_t
+ */
+int send_https_msg(char const *const host, char const *const port, char const *const api, const char *msg,
+                   const int msg_len);
+#endif /* end of include guard: UTILS_H_JKZACOAY */
```
Comment (jserv) [inline utils/utils.h]: What is `UTILS_H_JKZACOAY`?

Comment (splasky) [reply utils/utils.h]: This is automatically generate by vim snippets.

File: endpointComponent/endpoint.c [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+      response = uart_read(fd);
+      // real transmitted data
+      snprintf((char *)raw_msg, raw_msg_len, MSG, next_addr, response);
+      LE_DEBUG("Raw_msg:%s\n", raw_msg);
+      uint8_t private_key[AES_KEY_SIZE] = {0};
+      uint8_t id[IMSI_LEN + 1] = {0};
+
+      if (get_aes_key(private_key) != 0) {
+        LE_ERROR("%s\n", "get aes key error");
```
Comment (howjmay) [inline endpointComponent/endpoint.c]: Are we able to use a macro to wrapper `LOG_ERROR` to wrap `LE_ERROR`, since we may need to run ta-endpoint on other platform (i.e. testing on linux on laptop).
Maybe we could open another issue for this task

File: endpointComponent/endpoint.c (hunk 5 above)
Comment (howjmay) [reply endpointComponent/endpoint.c]: btw, Where is `HTTP_OK` defined?

File: utils/utils.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  fp = fopen(ADDR_LOG_PATH, "a");
+  if (!fp) {
+    perror("open addr_log.log failed:");
+    fclose(fp);
+    return -1;
+  }
+  snprintf(addr_log, 83, addr_log_template, next_addr);
+  fputs(addr_log, fp);
+  fclose(fp);
+  return 0;
```
Comment (howjmay) [inline utils/utils.c]: We should use some status code declared in enum instead of int here.

File: utils/crypto_utils.c [hunk 10]
```diff
@@ -134,16 +135,16 @@ int aes_decrypt(const unsigned char *ciphertext, int ciphertext_len, const unsig
    * Provide the message to be decrypted, and obtain the plaintext output.
    */
 
-  for (int i = 0; i < ciphertext_len; i += AES_BLOCK_SIZE) {
-    memset(buf, 0, AES_BLOCK_SIZE);
-    n = (ciphertext_len - i > AES_BLOCK_SIZE) ? 16 : (int)(ciphertext_len - i);
+  for (int i = 0; i < ciphertext_len; i += CBC_IV_SIZE) {
+    memset(buf, 0, CBC_IV_SIZE);
+    n = (ciphertext_len - i > CBC_IV_SIZE) ? 16 : (int)(ciphertext_len - i);
```
Comment (howjmay) [inline utils/crypto_utils.c]: Do we need to change the `16` here into `CBC_IV_SIZE`?

File: utils/crypto_utils.c [hunk 11]
```diff
@@ -93,16 +94,16 @@ int aes_encrypt(const unsigned char *plaintext, int plaintext_len, const unsigne
   /*
    * Encrypt plaintext
    */
-  for (int i = 0; i < new_len; i += AES_BLOCK_SIZE) {
-    memset(buf, 0, AES_BLOCK_SIZE);
-    n = (new_len - i > AES_BLOCK_SIZE) ? 16 : (int)(new_len - i);
+  for (int i = 0; i < new_len; i += CBC_IV_SIZE) {
+    memset(buf, 0, CBC_IV_SIZE);
+    n = (new_len - i > CBC_IV_SIZE) ? 16 : (int)(new_len - i);
```
Comment (howjmay) [inline utils/crypto_utils.c]: ditto

**Review by howjmay — CHANGES_REQUESTED** [review]

File: endpointComponent/endpoint.c (hunk 5 above)
Comment (splasky) [reply endpointComponent/endpoint.c]: Already defined inside defined_error.h.

File: endpointComponent/endpoint.c (hunk 8 above)
Comment (splasky) [reply endpointComponent/endpoint.c]: Create a new issue for these question?
Now we have lot of DEBUG or some macro flags during compiling for different platform and test target.

Comment (splasky) [reply endpointComponent/endpoint.c]: > Are we able to use a macro to wrapper `LOG_ERROR` to wrap `LE_ERROR`, since we may need to run ta-endpoint on other platform (i.e. testing on linux on laptop).
> Maybe we could open another issue for this task

Create new issue at #28

File: endpointComponent/endpoint.c (hunk 5 above)
Comment (howjmay) [reply endpointComponent/endpoint.c]: thx

File: endpointComponent/endpoint.c [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+#include "utils/serializer.h"
+#include "utils/tryte_byte_conv.h"
+#include "utils/uart_utils.h"
+#include "utils/utils.h"
+
+#define MSG "%s:%s"
+static int PORT = TA_PORT;
+static char HOST[1024] = TA_HOST;
+
+//--------------------------------------------------------------------------------------------------
```
Comment (jserv) [inline endpointComponent/endpoint.c]: Remove this line.

File: endpointComponent/endpoint.c [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ * Prints the on-line help and exits.  This function is called when "-h" or "--help" appears on
+ * the command-line or when the help command is invoked.
+ **/
+//--------------------------------------------------------------------------------------------------
```
Comment (jserv) [inline endpointComponent/endpoint.c]: Remove this line.

File: utils/crypto_utils.c [hunk 14]
```diff
@@ -29,7 +30,7 @@ int get_device_id(const char *device_id) {
     }
   }
 
-  strncpy((char *)device_id, imsi, IMSI_LEN);
+  strcpy((char *)device_id, imsi);
```
Comment (jserv) [inline utils/crypto_utils.c]: Why did you invoke `strcpy`?

File: utils/uart_utils.c [hunk 15]
```diff
@@ -81,11 +81,13 @@ char *uart_read(const int fd) {
 
   ssize_t rdlen = read(fd, buf, sizeof(buf) - 1);
   if (rdlen > 0) {
-    // printf("buf = %s\n", buf);
+#ifdef DEBUG
+    printf("buf = %s\n", buf);
```
Comment (jserv) [inline utils/uart_utils.c:85]: Is it necessary?

File: utils/crypto_utils.c (hunk 14 above)
Comment (splasky) [reply utils/crypto_utils.c]: After gcc version > 8 `strncpy` is for different usage. See `-Wstringop-truncation`

Comment (jserv) [reply utils/crypto_utils.c]: We need security advice. Can you show some reference materials?

File: utils/uart_utils.c (hunk 15 above)
Comment (splasky) [reply utils/uart_utils.c:85]: I think all of macro flags during compiling should be reduce. But I will not solve in this pull request. Should be do in refactor stage.

Comment (jserv) [reply utils/uart_utils.c:85]: Create a new issue for cleaning up existing codebase.

Comment (splasky) [reply utils/uart_utils.c:85]: New issue at #33

File: Makefile [hunk 16]
```diff
@@ -29,49 +29,67 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
...
 $(UTILS_OBJS):
 	$(MAKE) -C $(UTILS_PATH)
+
+## Build tests
 test: $(TEST_PATH) $(UTILS_OBJS)
 	$(MAKE) -C $(TEST_PATH)
 
-clean: clean_client clean_third_party clean_test
+## Build ta-endpoint legato app
+ta_legato_app:
```
Comment (jserv) [inline Makefile]: The target name is misleading. I would expect `legato` simply.

File: Makefile [hunk 17]
```diff
@@ -29,49 +29,67 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
...
+	# TODO: notify should make ta_lib with arm compiler first
+	mkapp -v -t wp77xx ta-endpoint.adef
```
Comment (jserv) [inline Makefile]: You shall detect Legato AF in advance.

Comment (splasky) [reply Makefile]: I have already tried this. But how can we detect current bash setting like after command `source legat_bashrc` environment inside makefile?

**Comment by jserv** [comment]
Use `git rebase -i` to rework existing git commits. I would like to functional changes rather than trivial patches.

File: Makefile (hunk 17 above)
Comment (jserv) [reply Makefile]: It sounds good. But, you shall improve the build instructions.

File: endpointComponent/Component.cdef [hunk 18]
```diff
@@ -0,0 +1,41 @@
+sources:
```
Comment (howjmay) [inline endpointComponent/Component.cdef]: Could we put it under a folder for different hardware instead putting under root directory

File: Makefile [hunk 19]
```diff
@@ -29,49 +29,66 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
...
 	$(MAKE) -C $(TEST_PATH)
 
-clean: clean_client clean_third_party clean_test
+## Build ta-endpoint legato app
+legato:
+	make clean && make ARM=y DEBUG=n $(TA_LIB) -j$(nproc) && mkapp -v -t wp77xx ta-endpoint.adef
+
+## Prints help for targets with comments
+help:
+	@cat $(MAKEFILE_LIST)|awk '/^##.*$$/{l1=$$0;getline;l2=(l1 "##" $$0); print l2 $$0}' | awk -F"##" '{split($$3,t,":");printf "\033[36m%-30s\033[0m %s\n",t[1],$$2}'
```
Comment (jserv) [inline Makefile]: Split into several pretty lines.

File: ta-endpoint.adef [hunk 20]
```diff
@@ -0,0 +1,12 @@
+executables:
+{
+    taEndpoint = (endpointComponent)
```
Comment (jserv) [inline ta-endpoint.adef]: Why? `endpointComponent` sounds weird. Can you simply use `component(s)`?

File: utils/utils.c [hunk 21]
```diff
@@ -0,0 +1,75 @@
+#include "utils.h"
```
Comment (jserv) [inline utils/utils.c]: Filename `utils/utils.c` is confusing. Can you use some functional names such `utils/url-rewriter.c`?

File: ta-endpoint.adef (hunk 20 above)
Comment (splasky) [reply ta-endpoint.adef]: We will have other components inside directory. The enpointComponent just for endpoint's  component. Change to `components` will be misleading. How about change to `endpointLegato`?

Comment (splasky) [reply ta-endpoint.adef]: see also:https://github.com/legatoproject/legato-af/tree/master/apps/sample/mqtt

Comment (jserv) [reply ta-endpoint.adef]: No camelCase allowed for directory name.

Comment (jserv) [reply ta-endpoint.adef]: or simply use `legato`.

File: utils/url_msg.h [hunk 22]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+#ifndef URL_MSG_H
+#define URL_MSG_H
```
Comment (jserv) [inline utils/url_msg.h]: `url_msg.h` sounds weird. Change it to `url.h`.

File: utils/url_msg.h [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <stdint.h>
+#define TA_HOST "tangle-accel.puyuma.org"
```
Comment (jserv) [inline utils/url_msg.h]: Don't hardcode hostname here.

File: utils/url_msg.h (hunk 22 above)
Comment (jserv) [reply utils/url_msg.h]: Also, add brief descriptions for this file.

File: legato/endpoint.c [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+const char *HOST = TA_HOST;
+
+/**
+ * Prints the on-line help and exits.  This function is called when "-h" or "--help" appears on
+ * the command-line or when the help command is invoked.
+ **/
+static void PrintHelp(void) {
+  puts(
+      "NAME\n"
+      "        ta-enpoint - Transcation information with tangle accerlator.\n"
```
Comment (jserv) [inline legato/endpoint.c]: Always mention "Tangle-accelerator`.

File: legato/endpoint.c [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+
+#define TA_HOST "tangle-accel.puyuma.org"
+#define TA_PORT "443"
+#define API "transaction/"
+#define SSL_SEED "nonce"
+#define ADDRESS                                                                \
+  "POWEREDBYTANGLEACCELERATOR999999999999999999999999999999999999999999999999" \
+  "999999A"
+#define ADDR_LEN 81
+#define ADDR_LOG_PATH "addr_log.log"
```
Comment (jserv) [inline legato/endpoint.c]: The path is confusing. You can rename it as `ta-endpoint.log`.

File: legato/endpoint.c [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+#include "connectivity/conn_http.h"
+#include "utils/crypto_utils.h"
+#include "utils/serializer.h"
+#include "utils/tryte_byte_conv.h"
+#include "utils/uart_utils.h"
+#include "utils/url.h"
+
+#define MSG "%s:%s"
+
+#define TA_HOST "tangle-accel.puyuma.org"
```
Comment (jserv) [inline legato/endpoint.c]: Don't hardcoded hostname.

File: main.c [hunk 27]
```diff
@@ -1,95 +1,37 @@
...
 
-#define HOST "tangle-accel.puyuma.org"
-#define PORT "443"
+#ifndef DEBUG
+#define MSG "%s:%s"
+#else
+#define MSG "%s:THISISMSG9THISISMSG9THISISMSG"
+#endif
+
+#define TA_HOST "tangle-accel.puyuma.org"
```
Comment (jserv) [inline main.c]: Don't hardcoded hostname.

Comment (jserv) [reply main.c]: You can set the default host via build system + macro.

File: utils/url.h [hunk 28]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+#ifndef URL_MSG_H
+#define URL_MSG_H
```
Comment (jserv) [inline utils/url.h]: Should be `UTILS_URL_H`.

File: endpointComponent/Component.cdef (hunk 18 above)
Comment (splasky) [reply endpointComponent/Component.cdef]: Related at #31

File: legato/Component.cdef [hunk 29]
```diff
@@ -0,0 +1,42 @@
+sources:
```
Comment (jserv) [inline legato/Component.cdef:1]: Use lower case `component.cdef`.

Comment (splasky) [reply legato/Component.cdef:1]: This is the default build name for legato build system.

File: utils/url.c [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+  free(http_res_body);
+  http_res_body = NULL;
+
+  ret = parser->status_code;
+  free(parser);
+  free(req);
+  return ret;
+}
+
+int log_address(const char *path, uint8_t *next_addr, size_t addr_len) {
```
Comment (jserv) [inline utils/url.c]: `log_address` is a misleading name. Ask @HowJMay for advice.

**Comment by jserv** [comment]
Update top-level `README.md` accordingly.

**Review by howjmay — CHANGES_REQUESTED** [review]
Add `close #20` in both git message and PR message

**Comment by jserv** [comment]
Improve the git commit messages. We expect expressive wording.

File: utils/url.h [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+#define RET_SUCCESS 0
+#define RET_FAILED 0xff
+
+/**
+ * @brief generate trytes from random
+ *
+ * @param len length of output
+ * @param out output array
+ */
+void gen_rand_trytes(uint16_t len, uint8_t *out);
```
Comment (jserv) [inline utils/url.h:27]: This is unrelated to URL. Move it into another header.

**Comment by howjmay** [comment]
We will integrate legato and ta-endpoint in another PR. The current one is too huge for taking future process, so I will close it.
